### PR TITLE
APP-87: Integrate configuration demos for different platforms into single demo

### PR DESCRIPTION
**Pull request by RostakaGmfun**
/cc @forGGe @rasendubi

### CODE REVIEW

File: common/c/targets/esp8266/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+add_library(target_support
+            STATIC
+            target.c
+            driver/uart.c
+            driver/gpio.c
+            )
+
+# resolve cyclic dependencies by forcing linker to walk through link libraries 3 times
+set_target_properties(target_support PROPERTIES LINK_INTERFACE_MULTIPLICITY 10)
```
Comment (forGGe) [inline common/c/targets/esp8266/CMakeLists.txt]: Misleading comment

File: common/c/targets/esp8266/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+# specify linker script
+configure_file(${CMAKE_CURRENT_SOURCE_DIR}/ld/eagle.rom.addr.v6.ld ${CMAKE_BINARY_DIR})
+configure_file(${CMAKE_CURRENT_SOURCE_DIR}/ld/eagle.app.v6.ld ${CMAKE_BINARY_DIR})
+set_target_properties(target_support PROPERTIES LINK_FLAGS "-T${CMAKE_BINARY_DIR}/eagle.app.v6.ld")
```
Comment (forGGe) [inline common/c/targets/esp8266/CMakeLists.txt]: Already set below. This line is no longer needed.

File: common/c/targets/esp8266/driver/gpio.c [hunk 3]
```diff
@@ -0,0 +1,207 @@
+/*
+ *  Copyright (C) 2014 -2016  Espressif System
+ *
+ */
+
```
Comment (forGGe) [inline common/c/targets/esp8266/driver/gpio.c:5]: Will not pass `rats`. Either add license header or add this file to excludes in maven

File: common/c/targets/x86-64/target.h [hunk 4]
```diff
@@ -28,17 +28,13 @@
...
-#include <string.h>
-#include <time.h>
 
 /* Demo print routine. Default printf exists for x86_64 target. */
 #define demo_printf(msg, ...) printf((msg), ##__VA_ARGS__)
 
 /* Initialises a target. Zero value means success, negative - errors. */
-static inline int target_initialise(void)
-{
+static inline int target_initialize(void) {
```
Comment (forGGe) [inline common/c/targets/x86-64/target.h]: Improper brace placement.

File: configurationdemo/source/c/CMakeLists.txt [hunk 5]
```diff
@@ -33,6 +33,8 @@ add_subdirectory(libs/kaa)
 # Directory containing target support library.
 add_subdirectory(targets/${KAA_TARGET})
 
+include_directories(targets/${KAA_TARGET})
```
Comment (forGGe) [inline configurationdemo/source/c/CMakeLists.txt]: If target_include_directories() executed with correct include dirs then this line is not required.

File: pom.xml [hunk 6]
```diff
@@ -268,16 +268,10 @@
                         <exclude>**/*xcodeproj/**</exclude>
                         <exclude>**/*.plist</exclude>
                         <exclude>**/*.cproject</exclude>
-                        <exclude>**/source/esp8266/ld/*.ld</exclude>
-                        <exclude>**/source/esp8266/driver/gpio.c</exclude>
-                        <exclude>**/source/esp8266/driver/uart.c</exclude>
-                        <exclude>**/source/esp8266/include/gpio*.h</exclude>
-                        <exclude>**/source/esp8266/include/driver/uart*.h</exclude>
+                        <exclude>**/common/c/targets/esp8266/ld/*.ld</exclude>
+                        <exclude>**/common/c/targets/esp8266/driver/*</exclude>
```
Comment (rasendubi) [inline pom.xml]: Don't use globbing here. Specify file name precisely.

File: common/c/targets/esp8266/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+cmake_minimum_required(VERSION 2.8)
+
+if(DEFINED $ENV{ESPRESSIF_HOME})
+    set(ESPRESSIF_HOME $ENV{ESPRESSIF_HOME})
+else()
+    set(ESPRESSIF_HOME /opt/Espressif)
+endif()
+
+set(ESP_SDK_HOME ${ESPRESSIF_HOME}/esp-rtos-sdk)
+set(ESP_TOOLCHAIN_LIBS_HOME ${ESPRESSIF_HOME}/crosstool-NG/builds/xtensa-lx106-elf/lib/gcc/xtensa-lx106-elf/4.8.2)
```
Comment (rasendubi) [inline common/c/targets/esp8266/CMakeLists.txt]: Can you change this path or make it configurable? I use properly installed toolchain, so I don't have `crosstool-NG/builds/xtensa-lx106-elf/lib/gcc` path. Check `toolchains/esp8266.cmake` in the SDK.

**Comment by RostakaGmfun** [comment]
Fixed and verified, re-review please
/cc @rasendubi

File: pom.xml [hunk 8]
```diff
@@ -268,16 +268,14 @@
...
-                        <exclude>**/source/esp8266/driver/gpio.c</exclude>
-                        <exclude>**/source/esp8266/driver/uart.c</exclude>
-                        <exclude>**/source/esp8266/include/gpio*.h</exclude>
-                        <exclude>**/source/esp8266/include/driver/uart*.h</exclude>
+                        <exclude>**/common/c/targets/esp8266/ld/*.ld</exclude>
+                        <exclude>**/common/c/targets/esp8266/driver/uart.c</exclude>
+                        <exclude>**/common/c/targets/esp8266/driver/uart.h</exclude>
+                        <exclude>**/common/c/targets/esp8266/driver/uart_register.h</exclude>
+                        <exclude>**/common/c/targets/esp8266/driver/gpio.c</exclude>
+                        <exclude>**/common/c/targets/esp8266/driver/gpio.h</exclude>
```
Comment (rasendubi) [inline pom.xml]: Don't use globbing. Specify exclude files explicitly.

**Comment by forGGe** [comment]
One comment pending

**Comment by forGGe** [comment]
Branch with merge conflicts, please fix.

File: common/c/targets/esp8266/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+else()
+    set(ESPRESSIF_HOME /opt/Espressif)
+endif()
+
+set(ESP_SDK_HOME ${ESPRESSIF_HOME}/esp-rtos-sdk)
+
+if(DEFINED ENV{ESP8266_TOOLCHAIN_PATH})
+  set(TOOLCHAIN_PATH "$ENV{ESP8266_TOOLCHAIN_PATH}")
+  message(STATUS "Toolchain path is provided: ${TOOLCHAIN_PATH}")
+else ()
```
Comment (forGGe) [inline common/c/targets/esp8266/CMakeLists.txt]: Extra space

File: common/c/targets/esp8266/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+  if (NOT TOOLCHAIN_PATH)
+    # Set default path.
+    set(TOOLCHAIN_PATH /opt/Espressif/crosstool-NG/builds/xtensa-lx106-elf)
+    message(STATUS "Default toolchain path will be used")
+  endif ()
```
Comment (forGGe) [inline common/c/targets/esp8266/CMakeLists.txt]: Same

File: common/c/targets/esp8266/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+endif()
+
+set(ESP_SDK_HOME ${ESPRESSIF_HOME}/esp-rtos-sdk)
+
+if(DEFINED ENV{ESP8266_TOOLCHAIN_PATH})
+  set(TOOLCHAIN_PATH "$ENV{ESP8266_TOOLCHAIN_PATH}")
+  message(STATUS "Toolchain path is provided: ${TOOLCHAIN_PATH}")
+else ()
+
+  if (NOT TOOLCHAIN_PATH)
```
Comment (forGGe) [inline common/c/targets/esp8266/CMakeLists.txt]: Same

File: common/c/targets/esp8266/driver/gpio.c (hunk 3 above)
Comment (forGGe) [inline common/c/targets/esp8266/driver/gpio.c:5]: Comment no longer valid: addressed by adding this file to exclude.

**Comment by forGGe** [comment]
Looks good

File: common/c/targets/esp8266/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+                      ${ESP_TOOLCHAIN_LIBS_HOME}/libgcc.a
+                      -T${CMAKE_CURRENT_SOURCE_DIR}/ld/eagle.app.v6.ld
+                      )
+
+message(STATUS "WiFi AP: ${WIFI_SSID}")
+message(STATUS "WiFi Pass: ${WIFI_PASSWORD}")
+
+# Expose WiFi credentials
+target_compile_definitions(target_support
+        PUBLIC
```
Comment (GamovCoder) [inline common/c/targets/esp8266/CMakeLists.txt]: indentation

File: common/c/targets/esp8266/target.c [hunk 13]
```diff
@@ -84,44 +99,46 @@ bool wifi_connect(const char *ssid, const char *pwd) {
...
+
+    wifi_print_station_config();
+    printf("Connecting to %s...\r\n ", ssid);
+    if (!wifi_station_connect()) {
+        printf("FAIL!\r\n");
+        return false;
+    }
+
+    uint8 status = wifi_station_get_connect_status();
+    while (status==STATION_CONNECTING) {
```
Comment (ohamov) [inline common/c/targets/esp8266/target.c]: cs

Comment (rasendubi) [inline common/c/targets/esp8266/target.c]: Rewrite as:

``` c
uint8 status;
do {
    status = wifi_station_get_connect_status();
} while (status == STATION_CONNECTING);

switch (status) {
    ...
}
```

That will be much easier to understand.

**Comment by forGGe** [comment]
looks good

**Comment by ohamov** [comment]
LGTM

**Comment by rasendubi** [comment]
Well, my comment wasn't addressed.

**Pull request merged**